### DISCUSSION TITLE
How to mark/reject resources using unsupported features

### DISCUSSION THREAD

**Discussion by mlavacca** [kubernetes-sigs/gateway-api#2617] [Ideas]
**Problem statement**:

At the moment there is no standardized way to check a `HTTPRoute` uses an unsupported feature. At Kong, we have a validating webhook that rejects HTTPRoutes using unsupported features. [GEP-2162](https://github.com/kubernetes-sigs/gateway-api/blob/main/geps/gep-2162.md) introduced the `supportedFeatures` field to highlight the features supported by the implementation, and in the [future work](https://github.com/kubernetes-sigs/gateway-api/blob/main/geps/gep-2162.md#research-the-development-of-an-unsupported-feature-warningblocking-mechanism) section such topic is mentioned. A standardized and shared way of accepting or rejecting resources using unsupported features is desirable, and this issue is about sensing the community's feelings.

**Proposed solution**

I think a viable (and desirable) solution would be to move this logic away from implementations' webhooks and instead rely on the resources' status. We could introduce a new reason for gateways and routes `Accepted` condition so that no resource using an unsupported feature is accepted and proper feedback is given to the user.

**Comment by pleshakov** [comment]
Could https://github.com/kubernetes-sigs/gateway-api/blob/a919f8b2b21ffb3aecae0a0a03b6e7ed0d17e95e/apis/v1/shared_types.go#L344 work? 

We use it in [our](https://github.com/nginxinc/nginx-gateway-fabric/tree/main) implementation for HTTPRoute fields that are not supported (ex. an unsupported filter). Although, we also use it for invalid values too

**Comment by mlavacca** [comment]
> Could
> 
> https://github.com/kubernetes-sigs/gateway-api/blob/a919f8b2b21ffb3aecae0a0a03b6e7ed0d17e95e/apis/v1/shared_types.go#L344
> 
> work?
> We use it in [our](https://github.com/nginxinc/nginx-gateway-fabric/tree/main) implementation for HTTPRoute fields that are not supported (ex. an unsupported filter). Although, we also use it for invalid values too

They are intended for a different use, when a field value does not belong to the enum. Example:
https://github.com/kubernetes-sigs/gateway-api/blob/6b8c23701bc03040a59771f17d73300fcea742a9/apis/v1/httproute_types.go#L318-L338

**Comment by mlavacca** [comment]
@shaneutt @robscott @youngnick any thoughts on this one? Is this something we could consider to implement?

**Reply by shaneutt** [reply]
The proposal to standardize on rejecting routes based on the supportedFeatures in GatewayClass, and providing some new relevant conditions/reasons to explain when things wont work sounds good to me :+1:

**Reply by youngnick** [reply]
Yes, I agree having standard conditions or reasons to explain that something is rejected because it uses an unsupported feature seems like a good idea to me as well.

**Reply by mlavacca** [reply]
Sounds good, I'll open a PR to update [GEP-2162](https://github.com/kubernetes-sigs/gateway-api/blob/main/geps/gep-2162.md) accordingly.

**Reply by pmalek** [reply]
For those looking for GEP-2162: it has been moved to https://github.com/kubernetes-sigs/gateway-api/blob/main/geps/gep-2162/index.md
